presto-main/src/main/java/com/facebook/presto/server/StatementResource.java

```diff
@@ -168,7 +170,14 @@ public Response createQuery(
     {
         assertRequest(!isNullOrEmpty(statement), "SQL statement is empty");
 
-        Session session = createSessionForRequest(servletRequest, transactionManager, accessControl, sessionPropertyManager, queryIdGenerator.createNextQueryId());
+        Session session;
+
+        try {
+            session = createSessionForRequest(servletRequest, transactionManager, accessControl, sessionPropertyManager, queryIdGenerator.createNextQueryId());
+        }
+        catch (PrestoException e) {
+            return Response.serverError().entity(createQueryError(e)).build();
```

Comment:
Can we use `toFailure(e).toFailureInfo()` instead of adding the `QueryErrorFactory` class?